### ISSUE TITLE
feat(embeddings): EmbeddingGemma (bidirectional Gemma 3 + Dense projections) and Qwen3-Embedding (last-token pooling) on the existing backbones

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1329]
## Summary

EmbeddingGemma (`model_type: gemma3_text`, `architectures: ["Gemma3TextModel"]`, `use_bidirectional_attention: true`) and Qwen3-Embedding (`model_type: qwen3`, `architectures: ["Qwen3ForCausalLM"]`, sentence-transformers layout with `1_Pooling`) are the two most downloaded small embedders and both run on decoder backbones mlxcel already has (`src/models/gemma3.rs`, `src/models/qwen3.rs`). Today both load as text generators. This issue adds two thin embedding wrappers: EmbeddingGemma runs the Gemma 3 layers with bidirectional masks (full layers: padding mask; sliding layers: bidirectional window), mean-pools, applies two bias-free Dense projections (768 -> 3072 -> 768) and normalizes; Qwen3-Embedding runs the causal Qwen3 layers with a causal+padding mask, takes the last real token and normalizes.

## Current behavior

- `src/models/detection.rs:184` routes `gemma3_text` to `ModelType::Gemma3`; `Gemma3Model::from_weights` (`src/models/gemma3.rs:898-926`) requires an `lm_head` (line 919) and ignores `dense.0.*` / `dense.1.*`; `forward_with_caches_and_embeddings` (line 737) always builds causal masks and ends with `self.lm_head.forward(&h)` (line 828).
- `src/models/detection.rs:168` routes `qwen3` to `ModelType::Qwen3`; `Qwen3Model::forward_impl` (`src/models/qwen3.rs:687-727`) ends with the lm head (or `embed_tokens.as_linear`) and exposes no hidden-state output.
- `Gemma3Model` hard-codes `window_size` into each `Attention` (`src/models/gemma3.rs:413`) and passes it to `attention_from_ptr` (line 188), so a sliding layer given an explicit mask still carries the causal window parameter.

### EmbeddingGemma

Config (verified on `mlx-community/embeddinggemma-300m-4bit`): `hidden_size 768`, `num_hidden_layers 24`, `num_attention_heads 3`, `num_key_value_heads 1`, `head_dim 256`, `intermediate_size 1152`, `query_pre_attn_scalar 256`, `rope_theta 1e6`, `rope_local_base_freq 1e4`, `sliding_window 512`, `_sliding_window_pattern 6` (`layer_types` lists five `sliding_attention` then one `full_attention`, repeating), `rms_norm_eps 1e-6`, `max_position_embeddings 2048`, `vocab_size 262144`, `pad_token_id 0`, `quantization {group_size 64, bits 4}` on the mlx-community conversion, `use_bidirectional_attention true`. The existing `gemma3::ModelArgs` (`src/models/gemma3.rs:47-88`) already parses every backbone key.

Forward (`B` rows, `L` tokens, `attention_mask` 1 = real):
```
h = embed_tokens[input_ids] * sqrt(hidden_size)                 # Gemma scaling, as get_embed_tokens
full_mask    = create_bidirectional_padding_mask(attention_mask)              # [B,1,1,L]
sliding_mask = create_bidirectional_window_mask(attention_mask, sliding_window) # |q-k| < 512 and not padding
for i, layer in layers:                                          # gemma3::TransformerBlock::forward
    is_global = (i + 1) % sliding_window_pattern == 0
    h = layer.forward(h, fresh KVCache (offset 0), Some(if is_global { full_mask } else { sliding_mask }))
h = norm(h)                                                      # model.norm, GemmaRMSNorm
pooled = mean_pool(h, attention_mask)                            # default mean; 1_Pooling absent on the mlx conversion
pooled = dense.1(dense.0(pooled))                                # 768 -> 3072 -> 768, no bias, no activation
embedding = normalize_l2(pooled)
```
For inputs up to 512 tokens the sliding mask equals the full mask; the window only matters between 512 and the 2048 cap.

Weight keys: the standard `model.embed_tokens.*`, `model.layers.{i}.*`, `model.norm.weight`, plus the projections. Two layouts exist:
- mlx-community conversion: `dense.0.{weight,scales,biases}` (logical `[3072, 768]`), `dense.1.{weight,scales,biases}` (`[768, 3072]`) in the main shards.
- sentence-transformers original: `2_Dense/model.safetensors` `linear.weight [3072, 768]` and `3_Dense/model.safetensors` `linear.weight [768, 3072]`, which `load_weights_from_dir_with_subfolders` exposes as `2_Dense.linear.weight` and `3_Dense.linear.weight`.

Sanitize: `{N}_Dense.linear.{weight,bias,scales,biases}` -> `dense.{k}.*` where `k` orders the folders by `N` ascending (assert `dense.0` has `out > in` and `dense.1` has `in > out`); drop `lm_head.*` and `head.*`; prefix bare `embed_tokens.` / `layers.` / `norm.` keys with `model.` when missing.

Limits: `max_length = min(sentence_bert_config max_seq_length (2048), model_max_length (2048), 8192) = 2048`. Prompts are caller-side: `task: search result | query: ` for queries, `title: none | text: ` for documents (document these in `docs/embeddings.md`). `dimensions` of 512, 256 and 128 are the trained Matryoshka sizes.

### Qwen3-Embedding

Config (verified on `Qwen/Qwen3-Embedding-0.6B`): standard Qwen3 (`hidden_size 1024`, 28 layers, 16/8 heads, `head_dim 128`, `rope_theta 1e6`, `tie_word_embeddings true`, `max_position_embeddings 32768`), `modules.json` with a Pooling module and `1_Pooling/config.json` `{pooling_mode_lasttoken: true, include_prompt: true}`. The tokenizer's post-processor appends `<|endoftext|>` (151643); `pad_token` is `<|endoftext|>`; `model_max_length 131072`, so `max_length = 8192` (the hard cap).

Forward:
```
h = embed_tokens[input_ids]
mask = create_causal_padding_mask(attention_mask, 0)             # [B,1,L,L]: causal and not padding
for i, layer in layers: h = layer.forward(h, fresh KVCache, Some(mask))   # qwen3::TransformerBlock::forward
h = norm(h)
embedding = normalize_l2(lasttoken_pool(h, attention_mask))      # position of the appended <|endoftext|>
```
Right padding is used (the last-token pooling handles it), which keeps the causal mask identical to the unpadded single-row case.

Weight keys: standard Qwen3; `lm_head.*` is dropped (tied). Sanitize adds the `model.` prefix when missing.

Query format is caller-side: `Instruct: {task}\nQuery: {query}`; documents are raw text. `dimensions` from 32 to 1024 are supported by the checkpoint.

## Implementation plan

1. `src/models/qwen3.rs`: split `forward_impl` into `pub(crate) fn forward_hidden(&self, input_ids, input_embeddings, caches, mask) -> UniquePtr<MlxArray>` (everything up to and including `norm`) and the head application; `forward_impl` calls both, byte-identical output. Add `// Used by: Qwen3 generation, Qwen3Embedding, #1356 (Qwen3 generative reranker)`.
2. `src/models/gemma3.rs`: make `Attention.window_size` settable after construction (it is already `pub`); no other backbone change.
3. `src/models/gemma3_embedding.rs` (+ `_tests.rs`): `Gemma3EmbeddingModel { embed_tokens: UnifiedEmbedding, layers: Vec<gemma3::TransformerBlock>, norm: GemmaRMSNorm, dense0, dense1: UnifiedLinear, hidden_size, sliding_window: i32, sliding_window_pattern, pooling, normalize }`. `load(dir)`: parse `gemma3::ModelArgs`, `load_weights_from_dir_with_subfolders`, sanitize, `UnifiedEmbedding::from_weights(.., "model.embed_tokens", gs, bits)`, `gemma3::TransformerBlock::from_weights(weights, &args, i)` for each layer then set `layer.self_attn.window_size = 0` (the embedding path owns its masks), `GemmaRMSNorm` from `model.norm.weight`, `UnifiedLinear::from_weights(.., "dense.0" / "dense.1", gs, bits)`. `impl EmbeddingModel` with the forward above; `default_pooling = Mean`, `embedding_dim = 768`. Per-layer caches: `gemma3::Cache::Standard(KVCache::new())` created per call and dropped.
4. `src/models/qwen3_embedding.rs` (+ `_tests.rs`): `Qwen3EmbeddingModel { model: Qwen3Model, pooling, normalize }`; `load(dir)` reuses `Qwen3Model::load` after verifying `lm_head` is absent or tied; `embed` builds `create_causal_padding_mask`, calls `model.forward_hidden(ids, None, &mut model.make_caches(), Some(&mask))`, pools with `lasttoken` (or the `1_Pooling` mode), normalizes. `default_pooling = LastToken`, `embedding_dim = hidden_size`.
5. `src/models/detection.rs`: inside `is_embedding_checkpoint` map `gemma3_text`/`gemma3` -> `ModelType::Gemma3Embedding` and `qwen3` -> `ModelType::Qwen3Embedding`; `src/model_metadata.rs` entries; `src/embeddings/loader.rs` arms.
6. `docs/supported-models.md` rows; `docs/embeddings.md` prompt prefixes and Matryoshka sizes for both.

## Validation

(a) Unit tests:
- `src/models/gemma3_embedding_tests.rs`: `dense_folder_keys_map_by_index` (both layouts produce `dense.0` `[3072,768]` and `dense.1` `[768,3072]`), `global_layers_use_padding_mask_and_sliding_layers_use_window` (2-layer synthetic config with pattern 2 and window 4: a token 6 positions away changes the global layer output but not the sliding one), `bidirectional_prefill_is_not_causal` (a 96-token synthetic input: changing the last token changes the first token's hidden state), `padding_invariance`.
- `src/models/qwen3_embedding_tests.rs`: `forward_hidden_then_head_matches_forward_impl` (guards the refactor, token-exact on a random 2-layer model), `causal_prefill_is_causal` (96 tokens, changing the last token leaves earlier hidden states unchanged within 1e-6), `last_token_pool_uses_appended_eos` (right-padded batch equals single-row result within 1e-4).

(b) Real checkpoints:
```bash
./target/release/mlxcel download mlx-community/embeddinggemma-300m-4bit
./target/release/mlxcel embed -m models/embeddinggemma-300m-4bit \
  -p "task: search result | query: Which planet is known as the Red Planet?" \
  -p "title: none | text: Mars, known for its reddish appearance, is often referred to as the Red Planet." \
  -p "title: none | text: Venus is the second planet from the Sun." \
  -p "title: none | text: Jupiter is the largest planet in our solar system."
./target/release/mlxcel-server -m models/embeddinggemma-300m-4bit --port 8080
curl -s localhost:8080/v1/embeddings -H 'Content-Type: application/json' \
  -d '{"input": ["task: search result | query: Which planet is known as the Red Planet?", "title: none | text: Mars, known for its reddish appearance, is often referred to as the Red Planet."], "dimensions": 256}'
```
Acceptance: 768-dim unit vectors; the Mars document scores highest for the query by at least 0.1 over the other two; `dimensions: 256` returns 256-dim unit vectors whose ranking is unchanged; a 1500-token document embeds without error and matches its unbatched result within 1e-3 (exercises the sliding window past 512).

```bash
./target/release/mlxcel download Qwen/Qwen3-Embedding-0.6B
./target/release/mlxcel embed -m models/Qwen3-Embedding-0.6B \
  -p "Instruct: Given a web search query, retrieve relevant passages that answer the query
Query: What is the capital of China?" \
  -p "Instruct: Given a web search query, retrieve relevant passages that answer the query
Query: Explain gravity" \
  -p "The capital of China is Beijing." \
  -p "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun."
```
Acceptance: the query-document similarity matrix equals `[[0.7646, 0.1414], [0.1355, 0.6000]]` within 2e-2 (the values in the checkpoint's model card for the bf16 original); identical texts give 1.0; a prompt of at least 64 tokens confirms causality (the bf16 run and an f16 run agree within 1e-2 and no NaN appears in a right-padded batch).

## Acceptance criteria

- [x] `Gemma3EmbeddingModel` loads both the mlx-community quantized layout and the `2_Dense`/`3_Dense` subfolder layout and runs bidirectional full layers plus bidirectional windowed sliding layers
- [x] `Qwen3EmbeddingModel` runs the causal backbone with a causal+padding mask and last-token pooling through the new `Qwen3Model::forward_hidden`
- [x] both checkpoints pass the gates above through `mlxcel embed` and `POST /v1/embeddings`
- [x] docs/supported-models.md updated
- [x] detection table in src/models/detection.rs updated with a test
- [x] cargo test --workspace --profile test-fast --features metal,accelerate passes
- [x] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass

Validation ran on Linux with GB10 and CUDA (`--profile test-fast --features cuda`), which is the equivalent of the macOS `metal,accelerate` gate on that host. The `2_Dense` / `3_Dense` subfolder layout is covered by a synthetic checkpoint written to disk and loaded through the real `Gemma3EmbeddingModel::load(dir)`, because only the mlx conversion of EmbeddingGemma is downloadable without accepting the gated `google/embeddinggemma-300m` terms.

## Out of scope

- Qwen3-Embedding-4B/8B (same code, only size).
- Server-side prompt-prefix injection (prefixes are caller-side; documented).

## Dependencies

Blocked by #1353

- Part of epic #1348
